File: content/snippets/find-duplicate-records/n1.test.py
import ast
import time
from pathlib import Path

import pytest
from sklearn.feature_extraction.text import TfidfVectorizer

import n0
from n1 import find_duplicates, normalise, record_text

LETTERS = "abcdefghijklmnopqrstuvwxyz"

# The same file as N0, with two pairs its blocking key cannot see: a name
# entered family-name first, and a postcode off by one digit.
CUSTOMERS = [
    {"name": "Jean Dupont", "city": "Paris", "postcode": "75011"},
    {"name": "Dupont Jean", "city": "Paris", "postcode": "75011"},
    {"name": "Marie Martin", "city": "Lyon", "postcode": "69003"},
    {"name": "Marie Martin", "city": "Lyon", "postcode": "69004"},
    {"name": "Paul Bernard", "city": "Bordeaux", "postcode": "33000"},
]

# Two records of one supplier, and a third company that merely spells like it.
SUPPLIERS = [
    {"name": "SNCF", "city": "Paris"},
    {"name": "Société Nationale des Chemins de Fer", "city": "Paris"},
    {"name": "SNEF", "city": "Paris"},
]



def scores(records):
    """Every pair and its score, threshold set aside."""
    return {(i, j): score for i, j, score in find_duplicates(records, threshold=0.0)}


# ---------------------------------------------------------------------------
# Point de rupture
# ---------------------------------------------------------------------------


def test_point_de_rupture_snef_obtient_un_meilleur_score_que_la_forme_developpee_de_sncf():
    """
    breaking_point : « « SNCF » et « Société Nationale des Chemins de Fer » sont
    une seule entreprise et ne partagent presque aucune tranche de lettres, tandis
    que « SNEF », qui en est une autre, obtient un meilleur score ».
    """
    pairs = scores(SUPPLIERS)
    assert pairs[(0, 2)] > pairs[(0, 1)]
    assert pairs[(0, 1)] < 0.2
    assert find_duplicates(SUPPLIERS) == []


def test_point_de_rupture_baisser_le_seuil_jusqu_au_vrai_doublon_fusionne_d_abord_les_deux_societes():
    """breaking_point : « baisser le seuil jusqu'à faire apparaître le vrai doublon fusionne d'abord les deux sociétés »."""
    pairs = scores(SUPPLIERS)
    loose = [(i, j) for i, j, _ in find_duplicates(SUPPLIERS, threshold=pairs[(0, 1)])]
    assert (0, 1) in loose and (0, 2) in loose
    between = [(i, j) for i, j, _ in find_duplicates(SUPPLIERS, threshold=(pairs[(0, 1)] + pairs[(0, 2)]) / 2)]
    assert between == [(0, 2)]


# ---------------------------------------------------------------------------
# Autres affirmations du niveau
# ---------------------------------------------------------------------------


def test_trouve_les_paires_que_la_cle_de_n0_ne_compare_jamais():
    """name : « TF-IDF sur n-grammes de caractères, voisins par cosinus » ; docstring : « Ce niveau abandonne la clé »."""
    assert [(i, j) for i, j, _ in find_duplicates(CUSTOMERS)] == [(0, 1), (2, 3)]
    assert n0.find_duplicates(CUSTOMERS, threshold=0.0) == []


def test_l_ordre_des_mots_ne_change_pas_le_score():
    """docstring : les n-grammes survivent « à un ordre des mots inversé »."""
    assert scores(CUSTOMERS)[(0, 1)] == 1.0


def test_une_faute_de_frappe_et_un_champ_tronque_gardent_un_score_au_dessus_du_seuil():
    """docstring : « ils survivent à une faute de frappe […] et à un champ tronqué »."""
    typo = [{"name": "Jean Dupont", "city": "Paris"}, {"name": "Jean Dupond", "city": "Paris"}]
    truncated = [{"name": "Société Générale", "city": "Paris"}, {"name": "Société Génér", "city": "Paris"}]
    assert find_duplicates(typo)[0][:2] == (0, 1)
    assert find_duplicates(truncated)[0][:2] == (0, 1)


def test_des_fiches_sans_rapport_ont_un_score_proche_de_zero():
    assert scores(CUSTOMERS)[(0, 4)] < 0.1


def test_les_n_grammes_restent_dans_les_mots():
    """docstring : « "dupont paris" borrows nothing from the join between the two words »."""
    analyser = TfidfVectorizer(analyzer="char_wb", ngram_range=(2, 4)).build_analyzer()
    assert not any(" " in gram.strip() for gram in analyser("dupont paris"))


def test_accents_et_casse_ne_sont_pas_une_difference():
    assert normalise("Société Générale") == "societe generale"
    pair = [{"name": "Société Générale", "city": "Paris"}, {"name": "SOCIETE GENERALE", "city": "paris"}]
    assert record_text(pair[0]) == record_text(pair[1])
    assert find_duplicates(pair)[0][2] == 1.0


def test_un_champ_tres_long_ne_noie_pas_le_score():
    note = "customer since 2019, prefers delivery in the afternoon, " * 20
    padded = [dict(record, note=note) for record in CUSTOMERS[:2]]
    assert find_duplicates(padded)[0][2] > 0.9


def test_rien_n_est_appris_ni_conserve_d_un_appel_a_l_autre():
    """docstring : « Rien ici n'est appris d'un corpus » ; regulatory : « rien du fichier n'est appris ni conservé au-delà de l'exécution »."""
    before = find_duplicates(CUSTOMERS)
    find_duplicates(SUPPLIERS * 50)
    assert find_duplicates(CUSTOMERS) == before


def test_sans_cle_toutes_les_paires_sont_examinees():
    """docstring : « sans clé, dans le pire des cas, toutes les paires » ; verdict : « les deux autres comparent, dans le pire des cas, toutes les paires »."""
    records = [{"name": f"Jean Dupont {i}", "city": "Paris"} for i in range(40)]
    assert len(find_duplicates(records, threshold=0.0)) == 40 * 39 // 2


def test_l_extrait_n_importe_que_scikit_learn_et_unicodedata():
    """risks.data_egress: none ; risks.deterministic: true."""
    source = ast.parse(Path(__file__).with_name("n1.py").read_text(encoding="utf-8"))
    imported = {a.name.split(".")[0] for n in ast.walk(source) if isinstance(n, ast.Import) for a in n.names}
    imported |= {n.module.split(".")[0] for n in ast.walk(source) if isinstance(n, ast.ImportFrom)}
    assert imported == {"unicodedata", "sklearn"}
    assert find_duplicates(CUSTOMERS) == find_duplicates(CUSTOMERS)


# ---------------------------------------------------------------------------
# Cas de production
# ---------------------------------------------------------------------------


def test_production_fichier_vide_et_fiche_unique():
    assert find_duplicates([]) == []
    assert find_duplicates([CUSTOMERS[0]]) == []


def test_production_des_fiches_sans_aucune_lettre_ne_font_pas_lever():
    assert find_duplicates([{"name": "", "city": ""}, {"name": "—", "city": "..."}]) == []


def test_production_deux_mille_fiches_dans_une_borne_large():
    records = [{"name": f"Client {i:05d} {LETTERS[i % 26]}{LETTERS[(i * 7) % 26]}", "city": "Paris"} for i in range(2000)]
    debut = time.perf_counter()
    find_duplicates(records)
    assert time.perf_counter() - debut < 30


def test_production_accents_decomposes_espaces_insecables_et_largeur_nulle():
    pair = [{"name": "Jean Dupont", "city": "Paris"}, {"name": "Jean\u00a0Dupo\u0302nt", "city": "PARIS"}]
    assert find_duplicates(pair)[0][2] == 1.0
    zero_width = [{"name": "Jean Dupont", "city": "Paris"}, {"name": "Jean Du\u200bpont", "city": "Paris"}]
    assert find_duplicates(zero_width)[0][:2] == (0, 1)


def test_production_au_seuil_zero_toutes_les_paires_sortent():
    assert len(find_duplicates(CUSTOMERS, threshold=0.0)) == 10


def test_production_des_fiches_identiques_sortent_au_seuil_un():
    assert find_duplicates([CUSTOMERS[0], dict(CUSTOMERS[0])], threshold=1.0) == [(0, 1, 1.0)]

File: content/snippets/find-duplicate-records/n1.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { findDuplicates, normalise, recordText } from './n1.js';
import * as n0 from './n0.js';

// The same file as N0, with two pairs its blocking key cannot see: a name
// entered family-name first, and a postcode off by one digit.
const CUSTOMERS = [
  { name: 'Jean Dupont', city: 'Paris', postcode: '75011' },
  { name: 'Dupont Jean', city: 'Paris', postcode: '75011' },
  { name: 'Marie Martin', city: 'Lyon', postcode: '69003' },
  { name: 'Marie Martin', city: 'Lyon', postcode: '69004' },
  { name: 'Paul Bernard', city: 'Bordeaux', postcode: '33000' },
];

// Two records of one supplier, and a third company that merely spells like it.
const SUPPLIERS = [
  { name: 'SNCF', city: 'Paris' },
  { name: 'Société Nationale des Chemins de Fer', city: 'Paris' },
  { name: 'SNEF', city: 'Paris' },
];

const indexesOf = (pairs) => pairs.map(([i, j]) => [i, j]);

/** Every pair and its score, threshold set aside. */
const scores = (records) => new Map(findDuplicates(records, 0).map(([i, j, s]) => [`${i},${j}`, s]));

const LETTERS = 'abcdefghijklmnopqrstuvwxyz';
const pairOf = (records, threshold) => indexesOf(findDuplicates(records, threshold));

// ---------------------------------------------------------------------------
// Point de rupture
// ---------------------------------------------------------------------------

test('point de rupture : « SNEF » obtient un meilleur score que la forme développée de « SNCF »', () => {
  const pairs = scores(SUPPLIERS);
  assert.ok(pairs.get('0,2') > pairs.get('0,1'));
  assert.ok(pairs.get('0,1') < 0.2);
  assert.deepEqual(findDuplicates(SUPPLIERS), []);
});

test('point de rupture : baisser le seuil jusqu’au vrai doublon fusionne d’abord les deux sociétés', () => {
  const pairs = scores(SUPPLIERS);
  const loose = pairOf(SUPPLIERS, pairs.get('0,1'));
  assert.ok(loose.some(([i, j]) => i === 0 && j === 1) && loose.some(([i, j]) => i === 0 && j === 2));
  assert.deepEqual(pairOf(SUPPLIERS, (pairs.get('0,1') + pairs.get('0,2')) / 2), [[0, 2]]);
});

// ---------------------------------------------------------------------------
// Autres affirmations du niveau
// ---------------------------------------------------------------------------

test('trouve les paires que la clé de N0 ne compare jamais', () => {
  assert.deepEqual(pairOf(CUSTOMERS), [[0, 1], [2, 3]]);
  assert.deepEqual(n0.findDuplicates(CUSTOMERS, 0), []);
});

test('l’ordre des mots ne change pas le score', () => {
  assert.equal(scores(CUSTOMERS).get('0,1'), 1);
});

test('une faute de frappe et un champ tronqué gardent un score au-dessus du seuil', () => {
  assert.deepEqual(pairOf([{ name: 'Jean Dupont', city: 'Paris' }, { name: 'Jean Dupond', city: 'Paris' }]), [[0, 1]]);
  assert.deepEqual(pairOf([{ name: 'Société Générale', city: 'Paris' }, { name: 'Société Génér', city: 'Paris' }]), [[0, 1]]);
});

test('des fiches sans rapport ont un score proche de zéro', () => {
  assert.ok(scores(CUSTOMERS).get('0,4') < 0.1);
});

test('l’IDF empêche une ville commune de rapprocher toutes les paires', () => {
  // Commentaire : « The inverse document frequency is what stops "paris" […]
  // from making every pair look alike ».
  const parisians = ['Jean Dupont', 'Marie Martin', 'Paul Bernard', 'Anne Moreau'].map((name) => ({ name, city: 'Paris' }));
  assert.ok(Math.max(...scores(parisians).values()) < 0.3);
});

test('accents et casse ne sont pas une différence', () => {
  assert.equal(normalise('Société Générale'), 'societe generale');
  const pair = [{ name: 'Société Générale', city: 'Paris' }, { name: 'SOCIETE GENERALE', city: 'paris' }];
  assert.equal(recordText(pair[0]), recordText(pair[1]));
  assert.equal(findDuplicates(pair)[0][2], 1);
});

test('un champ très long ne noie pas le score', () => {
  const note = 'customer since 2019, prefers delivery in the afternoon, '.repeat(20);
  const padded = CUSTOMERS.slice(0, 2).map((record) => ({ ...record, note }));
  assert.ok(findDuplicates(padded)[0][2] > 0.9);
});

test('rien n’est appris ni conservé d’un appel à l’autre', () => {
  const before = findDuplicates(CUSTOMERS);
  findDuplicates(Array(50).fill(SUPPLIERS).flat());
  assert.deepEqual(findDuplicates(CUSTOMERS), before);
});

test('sans clé, toutes les paires sont examinées', () => {
  const records = Array.from({ length: 40 }, (_, i) => ({ name: `Jean Dupont ${i}`, city: 'Paris' }));
  assert.equal(findDuplicates(records, 0).length, (40 * 39) / 2);
});

test('l’extrait n’importe rien et tient sans dépendance', () => {
  // docstring js : « it fits in one screen with no dependency » ; data_egress: none.
  const source = readFileSync(new URL('./n1.js', import.meta.url), 'utf8');
  assert.doesNotMatch(source, /^\s*import\s|\brequire\(|\bimport\(|\bfetch\(/m);
  assert.deepEqual(findDuplicates(CUSTOMERS), findDuplicates(CUSTOMERS));
});

// ---------------------------------------------------------------------------
// Cas de production
// ---------------------------------------------------------------------------

test('production : fichier vide et fiche unique', () => {
  assert.deepEqual(findDuplicates([]), []);
  assert.deepEqual(findDuplicates([CUSTOMERS[0]]), []);
});

test('production : des fiches sans aucune lettre ne font pas lever', () => {
  // Python lève ici (voir son test DÉFAUT).
  assert.deepEqual(findDuplicates([{ name: '', city: '' }, { name: '—', city: '...' }]), []);
});

test('production : deux mille fiches dans une borne large', () => {
  const records = Array.from({ length: 2000 }, (_, i) => ({ name: `Client ${String(i).padStart(5, '0')} ${LETTERS[i % 26]}${LETTERS[(i * 7) % 26]}`, city: 'Paris' }));
  const debut = performance.now();
  findDuplicates(records);
  assert.ok(performance.now() - debut < 30_000);
});

test('production : accents décomposés, espaces insécables et largeur nulle', () => {
  assert.equal(findDuplicates([{ name: 'Jean Dupont', city: 'Paris' }, { name: 'Jean\u00a0Dupo\u0302nt', city: 'PARIS' }])[0][2], 1);
  assert.deepEqual(pairOf([{ name: 'Jean Dupont', city: 'Paris' }, { name: 'Jean Du\u200bpont', city: 'Paris' }]), [[0, 1]]);
});

test('production : au seuil zéro, toutes les paires sortent', () => {
  assert.equal(findDuplicates(CUSTOMERS, 0).length, 10);
});

test('des fiches identiques ne sortent pas au seuil un', async () => {
  // Le cosinus de deux vecteurs identiques vaut 0,9999999999999999 par
  // arrondi flottant, sous le seuil ; Python les rend.
  assert.deepEqual(findDuplicates([CUSTOMERS[0], { ...CUSTOMERS[0] }], 1), [[0, 1, 1]]);
});
